File: intersection_nodes.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
// two pointer soution
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode ptr1 = headA;
        ListNode ptr2 = headB;
        int a = 0;
        int b = 0;
        while(ptr1 != null){
            ptr1 = ptr1.next;
            a++;
        }
        while(ptr2 != null){
            ptr2 = ptr2.next;
            b++;
        }
        ptr1 = headA; ptr2 = headB;
        while(a > b){
            ptr1 = ptr1.next;
            a--;
        }
        while(b > a){
            ptr2 = ptr2.next;
            b--;
        }
        while(ptr1 != ptr2){
            ptr1 = ptr1.next;
            ptr2 = ptr2.next;
        }
        return ptr1;
    }
}

/*
// using hash map O(n) time complexity O(n) space complexity
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA == null || headB == null)
            return null;
        if(headA == headB)
            return headA;
        HashMap<ListNode, Integer> map = new HashMap<>();
        while(headA != null){
            map.put(headA, headA.val);
            headA = headA.next;
        }
        while(headB != null){
            if(map.containsKey(headB))
                return new ListNode(map.get(headB));
            headB = headB.next;
        }
        return null;
    }
}
*/

File: intersectionNodes.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

'''
Time Complexity -->
O(m+n) since we are travesring and checking the length of the List A and List B

Space Compelxity -->
O(1)
'''

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        lena = 0
        lenb = 0
        curr1 = headA
        curr2 = headB
        while curr1:
            lena+=1
            curr1 = curr1.next
        while curr2:
            lenb+=1
            curr2 = curr2.next
        curr1 = headA
        curr2 = headB
        if lena>lenb:
            while lena!=lenb:
                curr1 = curr1.next
                lena-=1
        if lenb>lena:
            while lenb!=lena:
                curr2 = curr2.next
                lenb-=1
        
        while curr1 and curr2:
            if curr1 == curr2:
                return curr1
            curr1=curr1.next
            curr2=curr2.next
        return None
                
            
        